src/hotspot/share/opto/divnode.cpp

```diff
@@ -447,6 +447,73 @@ static Node *transform_long_divide( PhaseGVN *phase, Node *dividend, jlong divis
   return q;
 }
 
+template <typename TypeClass>
+Node* make_and(Node* a, Node* b);
+
+template <>
+Node* make_and<TypeLong>(Node* a, Node* b) {
+  return new AndLNode(a, b);
+}
+
+template <>
+Node* make_and<TypeInt>(Node* a, Node* b) {
+  return new AndINode(a, b);
+}
+
+template <typename TypeClass>
+Node* make_urshift(Node* a, Node* b);
+
+template <>
+Node* make_urshift<TypeLong>(Node* a, Node* b) {
+  return new URShiftLNode(a, b);
+}
+
+template <>
+Node* make_urshift<TypeInt>(Node* a, Node* b) {
+  return new URShiftINode(a, b);
+}
+
+template <typename TypeClass, typename Unsigned>
+Node* unsigned_div_ideal(PhaseGVN* phase, bool can_reshape, Node* div) {
+  // Check for dead control input
+  if (div->in(0) && div->remove_dead_region(phase, can_reshape)) {
+    return div;
+  }
+  // Don't bother trying to transform a dead node
+  if (div->in(0) && div->in(0)->is_top()) {
+    return nullptr;
+  }
+
+  const Type* t = phase->type(div->in(2));
+  const TypeClass* tl = t->cast<TypeClass>();
```

Comment:
I believe `is_int()` will assert when `t` is not a `TypeInt`, what you want here is a `try_cast` that calls `isa_int()` instead.

Reply:
If I understand correctly, you are referring to the fact that t might be top, which is not correctly handled here, right? If that is what you mean, I think I will fix it by handling this case separately like in unsigned_mod_ideal because `t` should only ever be TypeClass (i.e. long or int) or Top. Or am I missing something?

```
  const Type* t = phase->type(mod->in(2));
  if (t == Type::TOP) {
    return nullptr;
  }
  const TypeClass* ti = t->cast<TypeClass>();
```

Reply:
Yes, I saw you checking for `nullptr` in the following line and thought you confused `is_int()` with `isa_int()`, the former cannot return a `nullptr`.